src/datasource.ts

```diff
@@ -0,0 +1,300 @@
+        this.authToken = await this.refreshAuthTokenViaOAuth();
+      } else {
+        this.authToken = await this.refreshAuthTokenViaLocalAuth();
+      }
+    }
+    return { Authorization: `Bearer ${this.authToken!.token}` };
+  }
+
+  /**
+   * This is the typical way the plugin loads an access token when it's live (vs. local dev).
+   * We hit the /oauth/token API, passing the configured clientId and clientSecret.
+   * @returns the new AuthToken
+   */
+  private async refreshAuthTokenViaOAuth(): Promise<AuthToken> {
+    const env = this.instanceSettings.jsonData.criblOrgBaseUrl.endsWith('cribl-staging.cloud') ? 'staging' : 'production';
+    const response = await lastValueFrom(getBackendSrv().fetch<OAuthTokenResponse>({
+      method: 'POST',
+      url: this.proxiedUrl(`${env}Auth`),
+    }));
+    if (response.status === 200 && response.data?.access_token?.length > 0) {
+      return {
+        token: response.data.access_token,
+        expiresAt: Date.now() + (response.data.expires_in * 1000),
+      };
+    } else {
+      throw new Error(`Unexpected OAuth response: ${JSON.stringify(response, undefined, 2)}`);
+    }
+  }
+
+  /**
+   * For local dev, this uses the /auth/login API to get an auth token 
+   * @returns the new AuthToken token
+   */
+  private async refreshAuthTokenViaLocalAuth(): Promise<AuthToken> {
+    const response = await lastValueFrom(getBackendSrv().fetch<LocalLoginResponse>({
+      method: 'POST',
+      url: this.proxiedUrl('localAuth'),
+    }));
+    if (response.status !== 200) {
+      throw new Error('Failed to acquire auth token');
+    }
+    return {
+      token: response.data.token,
+      expiresAt: parseJwtExp(response.data.token) * 1000,
+    };
+  }
+
+  /**
+   * Used for proxying requests through the Grafana backend, which is required when sending sensitive data
+   * like clientId and clientSecret to the OAuth API.  We use "routes" defined in plugin.json.
+   * https://grafana.com/docs/grafana/latest/developers/plugins/add-authentication-for-data-source-plugins/#add-a-proxy-route-to-your-plugin
+   */
+  private proxiedUrl(routePath: string): string {
+    return `${this.instanceSettings.url}/${routePath}`;
+  }
+}
+
+/**
+ * Response we get back from /auth/login (for local dev)
+ */
+interface LocalLoginResponse {
+  token: string,
+}
+
+/**
+ * Response we get back from /oauth/token
+ */
+interface OAuthTokenResponse {
+  access_token: string,
+  scope: string,
+  expires_in: number,
+  token_type: 'Bearer',
+}
+
+/**
+ * Auth token, cached until it needs to be refreshed
+ */
+interface AuthToken {
+  token: string,
+  expiresAt: number,
+}
+
+/**
+ * Parse a JWT and get its expiration time
+ * @param token the encoded JWT
+ * @returns the expiration time in epoch seconds
+ */
+function parseJwtExp(token: string): number {
+  const b64 = token.split('.')[1].replace(/-/g, '+').replace(/_/g, '/');
```

Comment:
q: could we validate the JWT signature here? Is the server really the one we expected to talk to?